homeassistant/components/stream/hls.py

```diff
@@ -64,55 +78,184 @@ class HlsPlaylistView(StreamView):
     cors_allowed = True
 
     @staticmethod
-    def render_preamble(track):
-        """Render preamble."""
-        return [
-            "#EXT-X-VERSION:7",
-            f"#EXT-X-TARGETDURATION:{track.target_duration}",
-            '#EXT-X-MAP:URI="init.mp4"',
-        ]
+    def render(track, ll_hls):
+        """Render HLS playlist file."""
+        # NUM_PLAYLIST_SEGMENTS+1 because most recent is probably not yet complete
+        segments = list(track.get_segment())[-(NUM_PLAYLIST_SEGMENTS + 1) :]
 
-    @staticmethod
-    def render_playlist(track):
-        """Render playlist."""
-        segments = list(track.get_segment())[-NUM_PLAYLIST_SEGMENTS:]
+        # Create a copy of the last segment with the parts frozen for rendering playlist
+        segments[-1] = copy(segments[-1])
+        segments[-1].parts = segments[-1].parts.copy()
+
+        # Remove first segment if the last segment is actually complete
+        if segments[-1].duration:
+            segments = segments[-NUM_PLAYLIST_SEGMENTS:]
 
-        if not segments:
-            return []
+        target_duration = round(max(s.duration for s in segments if s.duration))
 
         playlist = [
+            "#EXTM3U",
+            "#EXT-X-VERSION:6",
+            "#EXT-X-INDEPENDENT-SEGMENTS",
+            '#EXT-X-MAP:URI="init.mp4"',
+            f"#EXT-X-TARGETDURATION:{target_duration}",
             f"#EXT-X-MEDIA-SEQUENCE:{segments[0].sequence}",
             f"#EXT-X-DISCONTINUITY-SEQUENCE:{segments[0].stream_id}",
+            "#EXT-X-PROGRAM-DATE-TIME:"
```

Comment:
From Appendix B.1:

>       All Media Playlists have EXT-PROGRAM-DATE-TIME tags.  This allows
      more-precise mapping between Segments across Renditions.  Note
      that real-time clocks are NOT required to be synchronized between
      client and server.